[Doc] Propose the RFC process (#4755)
taichi-dev/taichi: docs/rfcs/20220410-rfc-process.md

# RFC: The RFC (Request for Comments) Process

* Author(s): [Ye Kuang](https://github.com/k-ye)
* Date: 2022-04-10
* Relevant Issue: N/A

---

> # Summary

Reviewer: Should this be `TL;DR` according to the template file?

This doc proposes the standard way to make design decisions in Taichi. RFCs should be proposed and discussed before any major design or implementation gets landed in Taichi.

# Background

Throughout the development of Taichi, we often hear new developers complaining about the steep learning curve of Taichi's codebase. One reason is that, currently the design decisions are spread across different places and hard to track.

> RFC process is a well-established mechanism in the open source communities to solve this problem. It provdies a common place for developers to discuss design decisions publicly. It also serves as a doc library for people who want to learn about Taichi's architecture, evolving history, etc.

Reviewer: ```suggestion
RFC process is a well-established mechanism in the open source communities to solve this problem. It provides a common place for developers to discuss design decisions publicly. It also serves as a doc library for people who want to learn about Taichi's architecture, evolving history, etc.
```

# Goals

> * Provides a way for Taichi developers to discuss and to reach a consensus on any major changes openly and efficiently.

Author: Sure. I guess in practice, we will start with rough consensus..

* Serves as Taichi's public technical documentations.

> 

Reviewer 2: Cloud also mention that discussion on GitHub is also recommended before proposing the RFC?

Author: Done

# Detailed Design

1. Make a copy of [`yyyymmdd-rfc-template.md`](yyyymmdd-rfc-template.md), fill in all the required sections.
2. Send a new PR with this RFC file, then go through the normal PR process.

> 3. If the relevant community members decide to approve this RFC:

Reviewer 3: How can the author(s) determine who are the relevant members of the RFC here, when it comes to the case that the author(s) aren't internal devs?

Reviewer 2: I think a public Committer list should be fine for now.

   1. Open a new issue to track the work proposed by this RFC.
   2. Fill in the issue URL in `Relevant Issue`.
   3. Merge this PR & enjoy coding.
4. Otherwise if the members decide not to approve it, either re-iterate on the RFC and re-request the review, or simply drop it.

# Alternatives

## Other forms of documentation

RFC is not the only documentation system. We do have other forms of documentation, including [docsite](https://docs.taichi-lang.org/), technical blogs and design docs. Unlike RFCs, design docs will be written in a retrospective way to cover existing system's design and implementation.

> ## RFC ID

Author: Debatable.

Reviewer 4: I have no strong opinion on the solutions below. Using date SGTM.

Instead of using date, there are other options for the RFC ID:

> 1. Use an incremental number as the RFC ID. For example, [Rust](https://github.com/rust-lang/rfcs/blob/master/0000-template.md) has a four-digit ID system. However, we are very ambitious developers and feel like it is possible to overflow any given number of digits.

Reviewer 4: LOL!

# RFC: The RFC (Request for Comments) Process

* Author(s): [Ye Kuang](https://github.com/k-ye)
* Date: 2022-04-10
* Relevant Issue: N/A

---

# TL;DR

This doc proposes the standard way to make design decisions in Taichi.  The “RFC” (request for comments) process is intended to provide a consistent and controlled path for new features to land in Taichi, so that all stakeholders can be confident about the evolving direction.

# Background

Throughout the development of Taichi, we often hear new developers complaining about the steep learning curve of Taichi's codebase. One reason is that, currently the design decisions are spread across different places and hard to track.

RFC process is a well-established mechanism in the open source communities to solve this problem. It provides a common place for developers to discuss design decisions publicly. It also serves as a doc library for people who want to learn about Taichi's architecture, evolving history, etc.

# Goals

* Provides a way for Taichi developers to discuss and to reach a consensus on any major changes openly and efficiently.
* Serves as Taichi's public technical documentations.

# Detailed Design

> 1. Before initiating an RFC, it is recommended to file a Github Issue first to describe roughly the proposed solution.

Author: ```suggestion
1. Before initiating an RFC, it is recommended to file a Github Issue first to briefly describe the idea.
```

2. Make a copy of [`yyyymmdd-rfc-template.md`](yyyymmdd-rfc-template.md), fill in all the required sections.
3. Send a new PR with this RFC file, then go through the normal PR process.
4. If the relevant community members decide to approve this RFC:

>    1. In case you didn't do step 1, now is a good time to open a new issue to track the work proposed by this RFC.

Author: ```suggestion
   1. In case you haven't done step 1, now is a good time to open a new issue for tracking the works proposed by this RFC.
```
